server/debugsession.go

```diff
@@ -0,0 +1,834 @@
+package server
+
+import (
+	"bufio"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/google/go-dap"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime"
+	flowgo "github.com/onflow/flow-go/model/flow"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	sdk "github.com/onflow/flow-go-sdk"
+
+	emulator "github.com/onflow/flow-emulator"
+	"github.com/onflow/flow-emulator/server/backend"
+)
+
+type ScopeIdentifier uint
+
+const (
+	ScopeIdentifierLocal   ScopeIdentifier = 10000
+	ScopeIdentifierGlobal  ScopeIdentifier = 10001
+	ScopeIdentifierStorage ScopeIdentifier = 10002
+)
+
+type debugSession struct {
+	logger                *zerolog.Logger
+	backend               *backend.Backend
+	readWriter            *bufio.ReadWriter
+	variables             map[int]any
+	variableHandleCounter int
+	// sendQueue is used to capture messages from multiple request
+	// processing goroutines while writing them to the client connection
+	// from a single goroutine via sendFromQueue.
+	//
+	// We must keep track of the multiple channel senders with a wait group
+	// to make sure we do not close this channel prematurely.
+	//
+	// Closing this channel will signal the sendFromQueue goroutine that it can exit.
+	sendQueue chan dap.Message
+	sendWg    sync.WaitGroup
+	// debugger is the current
+	debugger          *interpreter.Debugger
+	stop              *interpreter.Stop
+	code              string
+	scriptLocation    common.StringLocation
+	scriptID          string
+	stopOnEntry       bool
+	configurationDone bool
+	launchRequested   bool
+	launched          bool
+	targetDepth       int
+}
+
+// sendFromQueue is to be run in a separate goroutine to listen
+// on a channel for messages to send back to the client.
+// It will return once the channel is closed.
+func (ds *debugSession) sendFromQueue() {
+	for message := range ds.sendQueue {
+		ds.logger.Trace().Msgf("DAP response: %#+v", message)
+
+		_ = dap.WriteProtocolMessage(ds.readWriter.Writer, message)
+		_ = ds.readWriter.Flush()
+	}
+}
+
+func (ds *debugSession) handleRequest() error {
+	request, err := dap.ReadProtocolMessage(ds.readWriter.Reader)
+	if err != nil {
+		return err
+	}
+
+	ds.logger.Trace().Msgf("DAP request: %#+v", request)
+
+	ds.sendWg.Add(1)
+	go func() {
+		ds.dispatchRequest(request)
+		ds.sendWg.Done()
+	}()
+
+	return nil
+}
+
+func (ds *debugSession) send(message dap.Message) {
+	ds.sendQueue <- message
+}
+
+func (ds *debugSession) dispatchRequest(request dap.Message) {
+	switch request := request.(type) {
+
+	case *dap.InitializeRequest:
```

Comment:
nit: I feel this could be extracted into functions and then just have a map of type to function call for different requests on the session struct.

Reply:
Yeah, the function become too big. 

Refactored in fdf0093445cf9b9fa18cb794730a5d108ad18434